File: xarm_moveit_demo/scripts/moveit_beeline_demo.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import rospy, sys
import moveit_commander
from moveit_commander import MoveGroupCommander
from geometry_msgs.msg import Pose
from copy import deepcopy
from geometry_msgs.msg import PoseStamped, Pose

class MoveItBeelineDemo:
    def __init__(self):

        moveit_commander.roscpp_initialize(sys.argv)
        rospy.init_node('moveit_beeline_demo', anonymous=True)

        # 初始化需要控制的规划组
        arm = MoveGroupCommander('xarm')

        # 允许重新规划
        arm.allow_replanning(True)

        # 设置位置(单位：米)和姿态（单位：弧度）的允许误差
        arm.set_goal_position_tolerance(0.005)
        arm.set_goal_orientation_tolerance(0.005)

        # 设置允许的最大速度
        arm.set_max_velocity_scaling_factor(0.8)

        # 控制机械臂先回到初始化位置
        arm.set_named_target('Home')
        arm.go()
        rospy.sleep(1)

        # 设置三角形第一个顶点的位姿，让机械臂从初始状态运动到这个点
        target_pose = PoseStamped()
        target_pose.header.frame_id = "base_link"
        target_pose.header.stamp = rospy.Time.now()
        target_pose.pose.position.x = 0.4
        target_pose.pose.position.y = 0.0
        target_pose.pose.position.z = 0.5
        target_pose.pose.orientation.w = 1

        # 设置机器臂当前的状态作为运动初始状态
        arm.set_start_state_to_current_state()

        # 设置机械臂终端运动的目标位姿
        arm.set_pose_target(target_pose)

        # 规划并执行
        traj = arm.go()
        rospy.sleep(1)

        # 获取当前位置并保存到start_pose里，可以用来作为路径点的起点和终点
        start_pose = arm.get_current_pose().pose
        end_pose = deepcopy(start_pose)

        # 初始化路点列表
        waypoints = []

        # 设置第一个路径点为start_pose
        wpose = deepcopy(start_pose)

        # 将start_pose加入路点列表
        waypoints.append(start_pose)

        # 设置第二个路径点, 把第二个路径点加入路点列表
        wpose.position.z -= 0.2
        waypoints.append(deepcopy(wpose))

        # 设置第三个路径点,把第三个路径点加入路点列表
        wpose.position.y += 0.2
        waypoints.append(deepcopy(wpose))

        # 把第四个路径点（end_pose,与第一个路径点重合）加入路点列表
        waypoints.append(deepcopy(end_pose))

        fraction = 0.0   # 路径规划覆盖率
        maxtries = 100   # 最大尝试规划次数
        attempts = 0     # 已经尝试规划次数

        # 设置机器臂当前的状态作为运动初始状态
        arm.set_start_state_to_current_state()

        # 尝试规划一条笛卡尔路径，依次通过所有路点
        while fraction < 1.0 and attempts < maxtries:
            (plan, fraction) = arm.compute_cartesian_path (
                                    waypoints,   # waypoint poses，路点列表，这里是4个点
                                    0.01,        # eef_step，终端步进值，每隔0.01m计算一次逆解判断能否可达
                                    0.0,         # jump_threshold，跳跃阈值，设置为0代表不允许跳跃
                                    True)        # avoid_collisions，避障规划
            # 尝试次数累加
            attempts += 1
            # 打印运动规划进程
            if attempts % 100 == 0:
                rospy.loginfo("Still trying after " + str(attempts) + " attempts...")

        # 如果路径规划成功（覆盖率100%）,则开始控制机械臂运动
        if fraction == 1.0:
            rospy.loginfo("Path computed successfully. Moving the arm.")
            arm.execute(plan)
            rospy.loginfo("Path execution complete.")
        # 如果路径规划失败，则打印失败信息
        else:
            rospy.loginfo("Path planning failed with only " + str(fraction) + " success after " + str(maxtries) + " attempts.")

        rospy.sleep(1)

        # 控制机械臂先回到初始化位置
        arm.set_named_target('Home')
        arm.go()
        rospy.sleep(1)

        moveit_commander.roscpp_shutdown()
        moveit_commander.os._exit(0)

if __name__ == "__main__":
    try:
        MoveItBeelineDemo()
    except rospy.ROSInterruptException:
        pass

File: xarm_moveit_demo/src/moveit_beeline_demo.cpp
#include <memory>
#include <rclcpp/rclcpp.hpp>
#include <moveit/move_group_interface/move_group_interface.h>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <thread>

/*
主要功能：
    1. 将机械臂移动到home位置
    2. 移动到三角形起点(0.4, 0.0, 0.5)
    3. 规划并执行三角形路径：
       - 向下移动20cm
       - 向右移动20cm
       - 返回起始点
    4. 最后返回home位置
 */

int main(int argc, char **argv)
{
    // 初始化ROS2节点
    rclcpp::init(argc, argv);
    auto const node = std::make_shared<rclcpp::Node>(
        "moveit_beeline_demo",
        rclcpp::NodeOptions().automatically_declare_parameters_from_overrides(true));

    // 创建异步执行器
    rclcpp::executors::SingleThreadedExecutor executor;
    executor.add_node(node);
    auto spinner = std::thread([&executor]()
                               { executor.spin(); });

    // 创建MoveGroup接口
    moveit::planning_interface::MoveGroupInterface arm(node, "xarm");

    // 设置规划参数
    arm.allowReplanning(true);            // 规划失败重新规划
    arm.setMaxVelocityScalingFactor(0.8); // 速度调节到最大速度的0.8

    // 移动到Home位置
    RCLCPP_INFO(node->get_logger(), "Moving to pose: Home");
    arm.setNamedTarget("Home");
    arm.move();

    // 设置三角形第一个顶点位姿
    RCLCPP_INFO(node->get_logger(), "Moving to pose: target_pose");
    geometry_msgs::msg::PoseStamped target_pose;
    target_pose.header.frame_id = "base_link";
    target_pose.header.stamp = node->now();
    target_pose.pose.position.x = 0.4;
    target_pose.pose.position.y = 0.0;
    target_pose.pose.position.z = 0.5;
    target_pose.pose.orientation.w = 1.0;

    arm.setStartStateToCurrentState();
    arm.setPoseTarget(target_pose);
    arm.move();

    // 获取并保存当前位置
    geometry_msgs::msg::Pose start_pose = arm.getCurrentPose().pose;
    geometry_msgs::msg::Pose end_pose = start_pose;

    // 定义三角形路径点
    std::vector<geometry_msgs::msg::Pose> waypoints;
    waypoints.push_back(start_pose);

    geometry_msgs::msg::Pose wppose = start_pose;
    wppose.position.z -= 0.2; // 向下移动20cm
    waypoints.push_back(wppose);

    wppose.position.y += 0.2; // 向右移动20cm
    waypoints.push_back(wppose);

    waypoints.push_back(end_pose); // 返回起点

    // 计算笛卡尔路径
    moveit_msgs::msg::RobotTrajectory trajectory; // 用于存储计算出的路径结果
    const double jump_threshold = 0.0;            // 跳跃阈值，用于检测路径中关节角度的突变
    const double eef_step = 0.01;                 // 末端执行器步长，表示路径中相邻路径点之间的最大距离，较小的点会生成更密集的路径点

    // 返回值成功规划的路径比例，1.0表示完全可以到达，0.8表示80%的路径可达剩余20%可能因碰撞或关节限制无法到达
    double fraction = arm.computeCartesianPath(
        waypoints, eef_step, jump_threshold, trajectory);

    RCLCPP_INFO(node->get_logger(),
                "Visualizing plan (Cartesian path) (%.2f%% achieved)", fraction * 100.0);

    // 执行路径
    if (fraction == 1.0)
    {
        arm.execute(trajectory);
    }

    // 返回Home位置
    RCLCPP_INFO(node->get_logger(), "Moving to pose: Home");
    arm.setNamedTarget("Home");
    arm.move();

    // 关闭节点
    rclcpp::shutdown();
    spinner.join();
    return 0;
}